Adding Seldon Go Operator for machine learning orchestration

axsauze opened the pull request:
This PR contains ongoing work to ensure we can add the Seldon Go operator into the community OperatorHub platform. There are currently changes that need to be added to address the issue as specified in #393. To reproduce issues, the instructions are in that github issue. We want to make sure the Seldon Go Operator aligns with the standards required for the community operators - please let me know what will be required to make sure this issue is addressed and we can make this operator available through this platform.

For context, [Seldon Core](https://www.seldon.io/open-source/) is an open source machine learning orchestration and deployment framework built on Kubernetes with integrations to popular ML frameworks (e.g. Tensorflow, Sklearn, etc).

axsauze commented:
I've addressed the issues in the automated builds. There are some complaints with the image tag (not sure why), but I would be keen to make sure that at least the fields added are correct, and the way that we've broken down our CRDs is aligned with what is required

anik120 commented:
/test CVP

axsauze commented:
Thank you for the prompt attention @anik120 - looking forward to exploring this further together with @robszumski and other relevant key redhat stakeholders.

aneeshkp commented:
@axsauze,  

seldonoperator.0.1.2.yaml is a crd file. Rename it to seldonoperator.crd.yaml(all CRD manifest filenames must have the suffix 'crd.yaml')
Make sure you have run operator-courier to validate the CSV
operator-courier verify --ui_validate_io seldon-operator/
ERROR:operatorcourier.validate:crd spec.group not defined.
ERROR:operatorcourier.validate:spec.icon can only contain two fields: "base64data" and "mediatype"
ERROR:operatorcourier.validate:UI validation failed to verify that required fields for operatorhub.io are properly formatted.
Resulting bundle is invalid, input yaml is improperly defined.

aneeshkp commented on upstream-community-operators/seldon-operator/seldonoperator.0.1.2.clusterserviceversion.yaml:
```diff
@@ -0,0 +1,267 @@
+spec:
+  replaces: seldonoperator.v1alpha1
+  displayName: Seldon Operator
+  description: |
```
Description can be more detailed, you can preview it in  https://operatorhub.io/preview

axsauze replied:
Perfect. I've added a more detailed definition.

axsauze commented on upstream-community-operators/seldon-operator/seldonoperator.0.1.2.clusterserviceversion.yaml:
```diff
@@ -0,0 +1,267 @@
+  maturity: alpha
+
+  icon:
+  - base64data: >- 
```
Thank you for the heads up. i've been able to use this to check that the email loads correctly.

axsauze commented on upstream-community-operators/seldon-operator/seldonoperator.0.1.2.clusterserviceversion.yaml:
```diff
@@ -0,0 +1,267 @@
+      group: seldon.io
+      kind: SeldonDeployment
+      version: 0.1.2
+      displayName: SeldonDelpoyment
```
Thank you, I have removed the duplicate.

axsauze commented:
I have addressed the issues provided. When I now run `operator-courier` it still gives me the following error:
```
ERROR: `CRD.spec.names.plural`.`CRD.spec.group` does not match CSV.spec.crd.owned.name [seldon-operator/seldonoperator.0.1.2.clusterserviceversion.yaml]
```

I have tried adding the group and plural keys into the CSV.spec.crd.owned.name object, which didn't help on removing this issue. I also tried adding the plural key under the names key to keep the structure of the CRD, but it still provides the same complain. 

Would you know what change would be required for this to work?

Thank you @aneeshkp, really appreciate the guidance.

aneeshkp commented:
> I have addressed the issues provided. When I now run `operator-courier` it still gives me the following error:
> 
> ```
> ERROR: `CRD.spec.names.plural`.`CRD.spec.group` does not match CSV.spec.crd.owned.name [seldon-operator/seldonoperator.0.1.2.clusterserviceversion.yaml]
> ```
> 
> I have tried adding the group and plural keys into the CSV.spec.crd.owned.name object, which didn't help on removing this issue. I also tried adding the plural key under the names key to keep the structure of the CRD, but it still provides the same complain.
> 
> Would you know what change would be required for this to work?
> 
> Thank you @aneeshkp, really appreciate the guidance.

In CRD file you have group=seldon.io and plural= plural: seldondeployments
So In your CSV.spec.crd.owned.name  should be "seldondeployments.seldon.io"
------------------------------------------------------------
customresourcedefinitions:
    owned:
    - kind: SeldonDeployment
      name: **seldondeployments.machinelearning.seldon.io**
----------------------------------------------------------------------
kind: CustomResourceDefinition
metadata:
  name: seldondeployments.machinelearning.seldon.io
spec:
  names:
    kind: SeldonDeployment
    listKind: SeldonDeploymentList
    plural: **seldondeployments**
    singular: seldondeployment
    shortNames:
    - sdep
  scope: Namespaced
  version: 0.1.2
  **group: seldon.io**

axsauze commented:
Ah that makes total sense, my bad! Thank you very much @aneeshkp - I have addressed those changes now, and it seems that there are no issues anymore running this controller. What would be the next steps on the testing phase? Thank you again.

aneeshkp commented on upstream-community-operators/seldon-operator/seldonoperator.0.1.2.clusterserviceversion.yaml:
```diff
@@ -0,0 +1,266 @@
+    certified: "false"
+    support: Clive Cox
+spec:
+  replaces: seldonoperator.v1alpha1
```
Replace operator not found, container fails to start

axsauze commented:
Thank you for the feedback @aneeshkp. We'll make sure we make the amenments and test - I created an issue (https://github.com/operator-framework/community-operators/issues/393) as we were attempting to run this and got a segfault in the scorecard. What did you use to run this and get that error? Was it this operator-sdk scorecard command? It would be get some guidance on this as I'm not sure whether there needs to be changes in our Go operator or just changes in the configuration.

aneeshkp commented:
> Thank you for the feedback @aneeshkp. We'll make sure we make the amenments and test - I created an issue (#393) as we were attempting to run this and got a segfault in the scorecard. What did you use to run this and get that error? Was it this operator-sdk scorecard command? It would be get some guidance on this as I'm not sure whether there needs to be changes in our Go operator or just changes in the configuration.

Follow this  document to manually deploy and validate 
https://github.com/operator-framework/community-operators/blob/master/docs/testing-operators.md

axsauze commented:
Thank you @aneeshkp, I have updated the file by removing the `replaces` as there is no operator it's currently replacing. I have tried following the tutorial you provided - I ran all the steps but I can't get the operator to successfully deploy. There does seem to be quite a lot of steps, some including pushing to quay.io (including using the helm plugin), so it seems there may have been an issue along the way. After debugging, I noticed that running the step `kubectl apply -f operator-source.yaml` results in issues with the following message:

```
NAME                        TYPE          ENDPOINT              REGISTRY          DISPLAYNAME   PUBLISHER   STATUS   MESSAGE                                                       AGE
alejandroseldon-operators   appregistry   https://quay.io/cnr   alejandroseldon                             Failed   The OperatorSource endpoint returned an empty manifest list   18s
```

I made sure to run all the services including OLM and operator marketplace, I have the application pushed into quay as `alejandroseldon/seldon-operator`, and I also made sure to have the token on the environment variable.

The contents of the `operator-source.yaml` file are as follows:
```
apiVersion: operators.coreos.com/v1
kind: OperatorSource
metadata:
  name: alejandroseldon-operators
  namespace: marketplace
spec:
  type: appregistry
  endpoint: https://quay.io/cnr
  registryNamespace: alejandroseldon
```

Would you know what the issue may be? Is there something that is not being done correctly during the tests? 

Thank you @aneeshkp.

aneeshkp commented:
@axsauze , Once you pushed to quay,  you should see it listed in your account's Applications tab.
If the application has a lock icon, click through to the application and its Settings tab and select to make the application public.
https://quay.io/application/ you and then click on application --> setting, it should be public.
Let me know if that worked for you.
https://github.com/operator-framework/community-operators/blob/master/docs/testing-operators.md

axsauze commented:
Thank you very much @aneeshkp - it seems the first time I did this I actually did that step but when trying to debug and re-create the quay.io application I forgot to make it public. I have now made it public, and have been able to run through all the healthchecks but I can't seem to find what the issue may be. Here is the results of the commands provided:

Here are all the containers I'm running (by running `kubectl get pods --all-namespaces -w`):
```
NAMESPACE     NAME                                            READY   STATUS    RESTARTS   AGE
kube-system   coredns-86c58d9df4-bz69v                        1/1     Running   2          39m
kube-system   coredns-86c58d9df4-j9bb5                        1/1     Running   2          39m
kube-system   etcd-docker-desktop                             1/1     Running   2          39m
kube-system   kube-apiserver-docker-desktop                   1/1     Running   2          39m
kube-system   kube-controller-manager-docker-desktop          1/1     Running   2          39m
kube-system   kube-proxy-cb2wb                                1/1     Running   2          39m
kube-system   kube-scheduler-docker-desktop                   1/1     Running   2          39m
marketplace   alejandroseldon-operators-6f4c89684b-crchk      1/1     Running   0          6m32s
marketplace   marketplace-operator-d8cc985d4-mn59s            1/1     Running   0          9m7s
marketplace   upstream-community-operators-68d8795fc9-dwdgr   1/1     Running   0          8m28s
olm           catalog-operator-dddbbb778-crdv2                1/1     Running   0          9m28s
olm           olm-operator-69c55565c8-mtdxv                   1/1     Running   0          9m28s
olm           olm-operators-j4sw9                             1/1     Running   0          9m22s
olm           operatorhubio-catalog-zsz45                     1/1     Running   0          9m22s
olm           packageserver-db76bb745-8fjnr                   1/1     Running   0          8m4s
olm           packageserver-db76bb745-s7pkb                   1/1     Running   0          8m16s
```

When I run `kubectl get operatorsource alejandroseldon-operators -n marketplace`:

```
NAME                        TYPE          ENDPOINT              REGISTRY          DISPLAYNAME   PUBLISHER   STATUS      MESSAGE                                       AGE
alejandroseldon-operators   appregistry   https://quay.io/cnr   alejandroseldon                             Succeeded   The object has been successfully reconciled   9m
```

When running `kubectl get catalogsourceconfig -n marketplace`:

```
NAME                           STATUS      MESSAGE                                       AGE
alejandroseldon-operators      Succeeded   The object has been successfully reconciled   10m
upstream-community-operators   Succeeded   The object has been successfully reconciled   11m
```

When running `kubectl get catalogsource -n olm`:

```
NAME                        NAME                  TYPE       PUBLISHER        AGE
alejandroseldon-operators   Custom                grpc       Custom           9m
olm-operators               OLM Operators         internal   Red Hat          12m
operatorhubio-catalog       Community Operators   grpc       OperatorHub.io   12
```

However when I run the `operator-subscription.yaml` and then `kubectl get clusterserviceversion -n default` I don't get any output:

```
No resources found.
```

The content of the files is as follows:

```
> cat operator-source.yaml
apiVersion: operators.coreos.com/v1
kind: OperatorSource
metadata:
  name: alejandroseldon-operators
  namespace: marketplace
spec:
  type: appregistry
  endpoint: https://quay.io/cnr
  registryNamespace: alejandroseldon

> cat catalog-source-config.yaml
apiVersion: operators.coreos.com/v1
kind: CatalogSourceConfig
metadata:
  name: alejandroseldon-operators
  namespace: marketplace
spec:
  targetNamespace: olm
  packages: seldon-operator

> cat operator-group.yaml
apiVersion: operators.coreos.com/v1alpha2
kind: OperatorGroup
metadata:
  name: machinelearning.seldon.io
  namespace: default
spec:
  targetNamespaces:
  - default

> cat operator-subscription.yaml
apiVersion: operators.coreos.com/v1alpha1
kind: Subscription
metadata:
  name: seldon-operator-subsription
  namespace: default
spec:
  channel: seldon-operator
  name: seldon-operator
  source: alejandroseldon-operators
  sourceNamespace: olm

```

Finally, for the channel I went into our seldon-core/helm-charts/seldon-core-operator folder where we have our seldon-operator helm chart, and ran the command `helm registry push --namespace alejandroseldon --channel seldon-operator quay.io`.

I did have to modify our `Chart.yaml` file to push, where I changed the name from `seldon-core-operator` to `seldon-operator` so that the command is able to push correctly to the repo.

Is there something in the configuration above that has not been done correctly or an assumption that was made that is incorrect? 

Thank you @aneeshkp.

aneeshkp commented:
@axsauze , the channel name is from your OLM bundle, package.yaml file. It should be alpha.
packageName: seldon-operator
channels: 
- name: alpha
  currentCSV: seldonoperator.0.1.2
defaultChannel: alpha

So change subscription.yaml  to 
apiVersion: operators.coreos.com/v1alpha1
kind: Subscription
metadata:
  name: seldon-operator-subsription
  namespace: default
spec:
  channel: alpha
  name: seldon-operator
  source: alejandroseldon-operators
  sourceNamespace: olm

But the deployment still failed. You can watch logs of catalog-operator for any errors.
Example: kubectl logs -f catalog-operator-7b8cd7f8bf-24rjt -n olm
Reason CustomResourceDefinition.apiextensions.k8s.io "seldondeployments.machinelearning.seldon.io" is invalid:
spec.versions[0].name: Invalid value:
Hope this help. Almost there .

 1 queueinformer_operator.go:186] Sync "default/install-h4585" failed: CustomResourceDefinition.apiextensions.k8s.io "seldondeployments.machinelearning.seldon.io" is invalid: [spec.versions[0].name: Invalid value: "0.1.2": a DNS-1035 label must consist of lower case alphanumeric characters or '-', start with an alphabetic character, and end with an alphanumeric character (e.g. 'my-name',  or 'abc-123', regex used for validation is '[a-z]([-a-z0-9]*[a-z0-9])?'), spec.version: Invalid value: "0.1.2": a DNS-1035 label must consist of lower case alphanumeric characters or '-', start with an alphabetic character, and end with an alphanumeric character (e.g. 'my-name',  or 'abc-123', regex used for validation is '[a-z]([-a-z0-9]*[a-z0-9])?')]
time="2019-06-05T03:26:46Z" level=info msg=syncing id=ujcOs ip=install-h4585 namespace=default phase=Failed
time="2019-06-05T03:26:46Z" level=info msg=syncing id=AtzcO ip=install-h4585 namespace=default phase=Failed

axsauze commented:
Thank you for the heads up on that, I have modified the group so that now the name is `seldondeployments.seldon` (group being `seldon`), but I still keep getting a healthcheck error when checking the logs of the catalog pod (logs below). 

As a heads up I also modified the contents of the `operator-group.yaml` to make sure it now matches the new name of the group (instead of `machinelearning.seldon.io`:
```
apiVersion: operators.coreos.com/v1alpha2
kind: OperatorGroup
metadata:
  name: seldon
  namespace: default
spec:
  targetNamespaces:
  - default
```

When I run `kubectl logs <CATALOG_POD> -n olm -f` I get the following errors:
```
time="2019-06-05T08:31:29Z" level=info msg="client hasn't yet become healthy, attempt a health check" currentSource="{olm-operators olm}" id=ueqC/ source=olm-operators
time="2019-06-05T08:31:30Z" level=info msg="building connection to registry" currentSource="{alejandroseldon-operators olm}" id=eYaRn source=alejandroseldon-operators
time="2019-06-05T08:31:30Z" level=info msg="client hasn't yet become healthy, attempt a health check" currentSource="{alejandroseldon-operators olm}" id=eYaRn source=alejandroseldon-operators
time="2019-06-05T08:31:30Z" level=info msg="building connection to registry" currentSource="{upstream-community-operators marketplace}" id=vcXEf source=upstream-community-operators
time="2019-06-05T08:31:30Z" level=info msg="client hasn't yet become healthy, attempt a health check" currentSource="{upstream-community-operators marketplace}" id=vcXEf source=upstream-community-operators
time="2019-06-05T08:31:31Z" level=info msg="building connection to registry" currentSource="{operatorhubio-catalog olm}" id=k6gdu source=operatorhubio-catalog
time="2019-06-05T08:31:31Z" level=info msg="client hasn't yet become healthy, attempt a health check" currentSource="{operatorhubio-catalog olm}" id=k6gdu source=operatorhubio-catalog
```

Do you know why the healthchecks seem to have an issue? Is this an issue with the CRD / CSV definition?

aneeshkp commented on upstream-community-operators/seldon-operator/seldonoperator.0.1.2.clusterserviceversion.yaml:
```diff
@@ -258,7 +258,7 @@ spec:
     owned:
     - kind: SeldonDeployment
-      name: seldondeployments.machinelearning.seldon.io
+      name: seldondeployments.seldon
```
seldondeployments.seldon" is invalid: [spec.group: Invalid value: "seldon": should be a domain with at least one dot.
metadata:name: Use as a name for the CustomResourceDefinition. It must use the {plural-name}.{group} format and use the values from the group and plural fields. If you do not follow these rules, you receive a validation error when installing the CRD.
Read here: https://github.com/kyma-project/community/blob/master/guidelines/custom-resource-definitions.md#location-and-file-name

axsauze replied:
thanks for the heads up @aneeshkp. I just want to confirm that I understand correctly as what I understood from what you had mentioned is that names that look like URLs are invalid, but below you advise to use our url. In that case should we change it back to the previous group - i.e. `seldondeployments.seldon`

aneeshkp replied:
Oh, I think previously the error was on the version, I thought it was label name too, sorry for the confusion.
You should see it while testing. here is the code that validates the crd's
https://github.com/operator-framework/operator-sdk/blob/master/vendor/k8s.io/apimachinery/pkg/util/validation/validation.go#L74

spec.version, you have 0.1.2, need to fix that.
The version names must contain the word alpha, for example, v1alpha1, if the software contains bugs, if enabling a feature can expose the bugs, and if a feature might be disabled by default.
or 
The version name must contain the word beta, for example, v2beta3, if the software is well-tested, enabling features is safe, has the features enabled by default, and support for the features is available, even though the details can change
or 
A stable definition must be versioned as vX where X is an integer, for example, v1, and it contains features which appear in multiple subsequent versions of the released software.
More information 
https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#alpha-beta-and-stable-versions

aneeshkp commented on upstream-community-operators/seldon-operator/seldonoperator.crd.yaml:
```diff
@@ -12,5 +12,5 @@ spec:
   scope: Namespaced
   version: 0.1.2
-  group: machinelearning.seldon.io
+  group: seldon
```
spec:group: The API group should reflect the collection of logically related objects. For example, all batch objects, such as Job or ScheduledJob, can belong to the batch API Group, such as batch.kyma-project.io. As best practice, use the fully-qualified domain name of the organization (kyma-project.io) preceded by a subgroup if necessary, for example crontab.kyma-project.io. The group name should reflect a capability-related and not an implementation-related name. For example, for eventing use eventing.kyma-project.io and not nats.kyma-project.io. Avoid prefixing the name with more than a subgroup, like in this example: user.crontab.kyma-project.io. If the subgroup consists of multiple words, do not use spaces, hyphens, or CamelCase.

aneeshkp commented on upstream-community-operators/seldon-operator/seldonoperator.crd.yaml:
```diff
@@ -1,7 +1,7 @@
 kind: CustomResourceDefinition
 metadata:
-  name: seldondeployments.machinelearning.seldon.io
+  name: seldondeployments.seldon
```
seldondeployments.seldon" is invalid: [spec.group: Invalid value: "seldon": should be a domain with at least one dot.
metadata:name: Use as a name for the CustomResourceDefinition. It must use the {plural-name}.{group} format and use the values from the group and plural fields. If you do not follow these rules, you receive a validation error when installing the CRD.
Read here: https://github.com/kyma-project/community/blob/master/guidelines/custom-resource-definitions.md#location-and-file-name

aneeshkp commented on upstream-community-operators/seldon-operator/seldonoperator.0.1.2.clusterserviceversion.yaml:
```diff
@@ -258,7 +258,7 @@ spec:
     - kind: SeldonDeployment
-      name: seldondeployments.machinelearning.seldon.io
+      name: seldondeployments.seldon
       version: 0.1.2
```
spec.versions[0].name: Invalid value: "0.1.2"
spec:version: Each API Group can exist in multiple versions. Use the version name in the URL, for example v1alpha1, v1beta1, or v1. For more details, see the Consideration section. For more information on versioning CRDs, see the Versioning section.
Read here: https://github.com/kyma-project/community/blob/master/guidelines/custom-resource-definitions.md#location-and-file-name

aneeshkp commented on upstream-community-operators/seldon-operator/seldonoperator.crd.yaml:
```diff
@@ -12,5 +12,5 @@ spec:
     - sdep
   scope: Namespaced
   version: 0.1.2
```
spec.versions[0].name: Invalid value: "0.1.2"
spec:version: Each API Group can exist in multiple versions. Use the version name in the URL, for example v1alpha1, v1beta1, or v1. For more details, see the Consideration section. For more information on versioning CRDs, see the Versioning section.
Read here: https://github.com/kyma-project/community/blob/master/guidelines/custom-resource-definitions.md#location-and-file-name

aneeshkp commented:
@axsauze any updates on the version fix?

axsauze commented:
Great @aneeshkp, it seems that the changes requested have made it work. I can now see that the seldon-operator is running correctly as follows:

When I run `kubectl logs <CATALOG_OPERATOR> -n olm -f`, I get:
```
time="2019-06-11T19:36:45Z" level=info msg=syncing id=pT2/j ip=install-pcnjb namespace=olm phase=Complete
time="2019-06-11T19:36:45Z" level=info msg=syncing id=yvyeE ip=install-77rgk namespace=default phase=Complete
```

However in order for the deployment to actually work, it's necessary that a secret is created for the operator, with the following yaml file:
```
apiVersion: v1
kind: Secret
metadata:
  name: seldon-operator-webhook-server-secret
  namespace: default
```

If this secret doesn't exist, then the operator would not work. Is there a way to make sure that the CRD is created together with this secret?

If this is the case then the operator should can be finalised to be part of the upstream community operators.

aneeshkp commented:
> However in order for the deployment to actually work, it's necessary that a secret is created for the operator, with the following yaml file:
> 
> apiVersion: v1
> kind: Secret
> metadata:
>   name: seldon-operator-webhook-server-secret
>   namespace: default
> If this secret doesn't exist, then the operator would not work. Is there a way to make sure that the CRD is created together with this secret?
> 
For now, you need to manually create the Secret. At some point in the future, we may allow them to be packaged/installed with an operator, though I’m not sure things like credentials should be stored in an operator package. In this case please mention prerequisites in the description so the admin knows to apply secrets for CSV to consume. Please update the description with details on how to install. Once that is done I will complete the manual testing from my side.

axsauze commented:
Perfect @aneeshkp. I have added the changes requested, including the updated version which reflects the requirements, as well as adding in the description the installation instructions. I have also added the operator to the community operators as well. I've tested it locally and it works as expected. Please let me know if this is all you need or something else needs to be modified.

aneeshkp commented:
> Perfect @aneeshkp. I have added the changes requested, including the updated version which reflects the requirements, as well as adding in the description the installation instructions. I have also added the operator to the community operators as well. I've tested it locally and it works as expected. Please let me know if this is all you need or something else needs to be modified.
@axsauze, community operator and upstream operator should be a separate PR. Please delete community operator form this PR and create a new one for community PR.

axsauze commented:
@aneeshkp Perfect, just removed it, now this PR only contains the upstream operator. I will submit a PR for the community operator once this is confirmed and merged. Thank you very much for your support.

axsauze commented:
@aneeshkp I didn't see your reply, I think you edited my comment above so didn't get notifications, but I have now actioned your request, it makes sense. I've had a look at the comments in the Travis CI logs, and it seems that it now succeeds and attempts to run the SeldonContainer. The error seems to be because the SeldonContainer doesn't start due to the secret not being created - this can be verified by running a `kubectl describe pods` command, and once the secret is created it should succeed (the pod may require being restarted if it's already in state ContainerCreating). Please let me know if any further changes are required.

aneeshkp commented:
@axsauze , Thanks for the changes , I was able to create pods successfully.
Now before I merge , Can you make the requirement of secret as its own section in description, So that it stands out 
You can see some examples here.
[https://operatorhub.io/operator/camel-k](https://operatorhub.io/operator/camel-k)
[https://operatorhub.io/operator/aqua](https://operatorhub.io/operator/aqua)

axsauze commented:
For sure - I've updated the description to reflect the secret, as well as an example model that can be deployed. Let me know what your thoughts are and I will be able to make any relevant changes.

aneeshkp commented:
@axsauze I am not sure how  I missed this , I tried to test in clean minikube and the seldon-operator-controller-manager log shows issue with cluster scope.
E0613 14:03:28.693026       1 reflector.go:134] sigs.k8s.io/controller-runtime/pkg/cache/internal/informers_map.go:126: Failed to list *v1.Service: services is forbidden: User "system:serviceaccount:default:default" cannot list resource "services" in API group "" at the cluster scope
E0613 14:03:28.694958       1 reflector.go:134] sigs.k8s.io/controller-runtime/pkg/cache/internal/informers_map.go:126: Failed to list *v1.Deployment: deployments.apps is forbidden: User "system:serviceaccount:default:default" cannot list resource "deployments" in API group "apps" at the cluster scope.   does this requires cluster permission?
E0613 14:03:29.677703       1 reflector.go:134] sigs.k8s.io/controller-runtime/pkg/cache/internal/informers_map.go:126: Failed to list *v1alpha2.SeldonDeployment: seldondeployments.machinelearning.seldon.io is forbidden: User "system:serviceaccount:default:default" cannot list resource "seldondeployments" in API group "machinelearning.seldon.io" at the cluster scope
E0613 14:03:29.694884       1 reflector.go:134] sigs.k8s.io/controller-runtime/pkg/cache/internal/informers_map.go:126: Failed to list *v1.Service: services is forbidden: User "system:serviceaccount:default:default" cannot list resource "services" in API group "" at the cluster scope
E0613 14:03:29.696226       1 reflector.go:134] sigs.k8s.io/controller-runtime/pkg/cache/internal/informers_map.go:126: Failed to list *v1.Deployment: deployments.apps is forbidden: User "system:serviceaccount:default:default" cannot list resource "deployments" in API group "apps" at the cluster scope

aneeshkp commented:
Also after testing the service account issue  please squash your commits
Reason : Contain more than one commit. **Please squash your commits.**

axsauze commented:
Thank you @aneeshkp. I am trying to understand, why did this issue suddenly arise? We are using the default service account, so not sure why it would fail like this. Did you create the secret before running it? Would you be able to tell me what you get when you run `kubectl get clusterrolebinding` and when you run `kubectl get secret`? Some pointers would help to know what needs to be changed. Thank you @aneeshkp, keen to get this over the line.

aneeshkp commented:
> Thank you @aneeshkp. I am trying to understand, why did this issue suddenly arise? We are using the default service account, so not sure why it would fail like this. Did you create the secret before running it? Would you be able to tell me what you get when you run `kubectl get clusterrolebinding` and when you run `kubectl get secret`? Some pointers would help to know what needs to be changed. Thank you @aneeshkp, keen to get this over the line.
@axsauze here  is the output. 
kubectl get secret
NAME                                    TYPE                                  DATA   AGE
default-token-l22nz                     kubernetes.io/service-account-token   3      7m46s
seldon-operator-webhook-server-secret   Opaque                                0      34s

and here is the clusterrolebinding out put 
[https://paste.fedoraproject.org/paste/0PYfK2G~I7GeJlIb5RYi-w](https://paste.fedoraproject.org/paste/0PYfK2G~I7GeJlIb5RYi-w)
Hope this helps.

axsauze commented:
Perfect, ok, so we should be able to see a service-account called "seldon-operator-manager-rolebinding", but it's not there. We have a file internally that we use to create this, which is as follows:
```
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  creationTimestamp: null
  name: seldon-operator-manager-rolebinding
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: seldon-operator-manager-role
subjects:
- kind: ServiceAccount
  name: default
  namespace: {{ .Release.Namespace }}
```

I am trying to understand how this could be added into the CSV, but the exampes I've been using don't define a service account. Is there any examples of CSVs that define it? Alternatively where would this be added?

aneeshkp commented:
> ClusterRoleBinding

Your InstallMode is single namespace/OwnNamespace. Why does it need Cluster permission?
 Creation of both permissions and clusterPermissions are now created via InstallPlans. If you have AllNameSpace then OLM will create cluster level permissions. But if you have specific RoleBinding references you should able to drop it OLM Bundle.
@dmesser @robszumski  any inputs here?

aneeshkp commented:
@axsauze  any updates from your side on why you need cluster level permission? if it requires cluster level permission then you need to setup install mode to AllNamespace.

axsauze commented:
@aneeshkp I was waiting for yourself or the people you tagged to could give us some pointers. I wasn't aware that the service account issue was directly related to cluster-wide permissions, but you are correct in that our operator does require AllNamespace, thanks for the heads up, I've have now updated that so the CSV file reflects that. Would that address the issue we were facing previously?

aneeshkp commented:
@axsauze , It works. Setting AllNamespace true adds cluster permission, Now  I am able to test and it works.
Can you add one more meta data  minKubeVersion:, Once this is done I can merge these changes.
MinKubeVersion: A minimum version of Kubernetes that server is supposed to have so operator(s) can be deployed. The Kubernetes version must be in "Major.Minor.Patch" format (e.g: 1.11.0).

https://github.com/operator-framework/operator-lifecycle-manager/blob/master/Documentation/design/building-your-csv.md

aneeshkp commented:
Also Please squash your commits
Reason : Contain more than one commit. Please squash your commits.

axsauze commented:
@aneeshkp the minkubeversion has been added under metadata, and the commits have been squashed, let me know if that works. One thing to note is that I had a look at the [cockroachdb example](https://github.com/operator-framework/community-operators/blob/82db6a3cadcddd815481897b385b51e75e89cd76/community-operators/cockroachdb/cockroachdb.v2.1.1.clusterserviceversion.yaml) and they seem to add the minkubeversion on the spec. Where should this key go? Is it correct where it's been added now?
